### PR TITLE
feat(api key management) Extract api key model

### PR DESCRIPTION
**Pull request by floganz**
## Context

One tiny step towards advanced API key management.

## Description

Introduces new model `ApiKey` that will replace `api_key` column on the `Organization` model during next steps.
Migrate existing values for `api_key` to the dedicated model.
Adjust REST API authorization to use lookup for ApiKey model and temporary keep previous lookup for Organization model.

### CODE REVIEW

File: app/controllers/api/base_controller.rb [hunk 1]
```diff
@@ -27,8 +27,9 @@ def authenticate
       true
     end
 
-    def current_organization(api_key = nil)
-      @current_organization ||= Organization.find_by(api_key:)
+    def current_organization(value = nil)
+      @current_organization ||=
+        ApiKey.find_by(value:)&.organization || Organization.find_by(api_key: value)
```
Comment (ancorcruz) [inline app/controllers/api/base_controller.rb:32]:
```suggestion
    def current_organization(api_key = nil)
      @current_organization ||=
        ApiKey.find_by(value: api_key)&.organization || Organization.find_by(api_key:)
```

I would keep `api_key` as name of the signature param as it express what it is at first glance without reading the implementation.

File: app/models/api_key.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    return generate_value if api_key.present?
+
+    value
+  end
+
+  private
+
+  def set_value
+    self.value = generate_value
+  end
```
Comment (ancorcruz) [inline app/models/api_key.rb:23]: I would avoid recursion in Ruby when possible and instead of `find_by` use `exists?` for better performance.

also the method could be a private one, is there any reason to expose it a public api of the model?

```suggestion
  before_create :set_key_value
  
  private

  def set_key_value
    begin
      self.value = SecureRandom.uuid 
    end while self.class.exists?(value: self.value)
  end
```

This should do the trick.

File: app/services/organizations/create_service.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+module Organizations
+  class CreateService < BaseService
+    def initialize(params)
+      @params = params
+      super
+    end
+
+    def call
+      organization = Organization.new(params)
```
Comment (ancorcruz) [inline app/services/organizations/create_service.rb:11]: would you avoid mass assignment here and list each attribute assigned to the organization like in the original code?

File: db/migrate/20241021140054_create_api_keys.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+class CreateApiKeys < ActiveRecord::Migration[7.1]
+  def up
+    create_table :api_keys, id: :uuid do |t|
+      t.references :organization, type: :uuid, null: false, foreign_key: true
+
+      t.string :value, null: false
```
Comment (ancorcruz) [inline db/migrate/20241021140054_create_api_keys.rb:8]: would you also add a unique index scoped to the organization_id?

File: db/migrate/20241021140054_create_api_keys.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      t.timestamps
+    end
+
+    safety_assured do
+      execute <<-SQL
+      INSERT INTO api_keys (value, organization_id, created_at, updated_at)
+      SELECT organizations.api_key, organizations.id, organizations.created_at, organizations.created_at
+      FROM organizations
+      SQL
+    end
```
Comment (ancorcruz) [inline db/migrate/20241021140054_create_api_keys.rb:19]: Do you also plan to drop the api_key column in organizations table at some point?

File: app/models/api_key.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+class ApiKey < ApplicationRecord
+  include PaperTrailTraceable
+
+  belongs_to :organization
+
```
Comment (ancorcruz) [inline app/models/api_key.rb:7]: I would also add a validation of value presence here? what do you think?

File: spec/factories/organizations.rb [hunk 7]
```diff
@@ -13,7 +13,11 @@
     end
 
     after(:create) do |organization, evaluator|
-      organization.webhook_endpoints.create(webhook_url: evaluator.webhook_url)
+      if evaluator.webhook_url
+        organization.webhook_endpoints.create!(webhook_url: evaluator.webhook_url)
+      end
+
+      organization.api_keys.create!
```
Comment (ancorcruz) [inline spec/factories/organizations.rb:20]: could you instead of creating the api_keys here add it to the organizations factory?

File: spec/models/api_key_spec.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    let(:unique_value) { SecureRandom.uuid }
+
+    before do
+      allow(SecureRandom).to receive(:uuid).and_return(used_value, unique_value)
+    end
+
+    it 'returns unique value between all ApiKeys' do
+      expect(subject).to eq unique_value
+    end
+  end
```
Comment (ancorcruz) [inline spec/models/api_key_spec.rb:60]: I don't see the need of testing these two methods (in my previous comment I have combined both into a single private method) as long as the `save` behavior is tested.

File: spec/models/api_key_spec.rb [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      end
+    end
+
+    context 'with a persisted record' do
+      let(:api_key) { create(:api_key) }
+
+      it 'does not call #set_value' do
+        expect(api_key).not_to have_received(:set_value)
+      end
+    end
```
Comment (ancorcruz) [inline spec/models/api_key_spec.rb:30]: I would avoid stubbing when possible...
here you can just check an uuid is assigned to a new records when saved,
it does not change when the record is already persisted and have one assigned...
and it handles the collision with an existing record with the same value (here you will have to stub SecureRandom I guess).

**Review by ancorcruz — CHANGES_REQUESTED** [review]
Good stuff, I have left you a few comments.

Let me know if you would like to have a chat about my comments.

File: app/services/users_service.rb [hunk 10]
```diff
@@ -29,7 +29,10 @@ def register(email, password, organization_name)
 
     ActiveRecord::Base.transaction do
       result.user = User.create!(email:, password:)
-      result.organization = Organization.create!(name: organization_name, document_numbering: 'per_organization')
+
+      result.organization = Organizations::CreateService
+        .call(name: organization_name, document_numbering: 'per_organization')
+        .organization
```
Comment (vincent-pochet) [inline app/services/users_service.rb:35]: To raise the new service error (just in case):

```suggestion
      result.organization = Organizations::CreateService
        .call(name: organization_name, document_numbering: 'per_organization')
        .raise_if_error!
        .organization
```

File: db/migrate/20241021140054_create_api_keys.rb (hunk 4 above)
Comment (vincent-pochet) [reply db/migrate/20241021140054_create_api_keys.rb:8]: :+1: 

An index on `organization_id` is also mandatory since we are doing `organization.api_keys.first` in multiple places

File: spec/models/organization_spec.rb [hunk 11]
```diff
@@ -138,11 +139,43 @@
...
 
-  describe 'Callbacks' do
-    it 'generates the api key' do
-      organization.save!
+  describe '#save' do
+    subject { organization.save! }
 
-      expect(organization.api_key).to be_present
+    before do
+      allow(organization).to receive(:generate_document_number_prefix).and_call_original # rubocop:disable RSpec/SubjectStub
```
Comment (vincent-pochet) [inline spec/models/organization_spec.rb:146]: Stubbing private method is not a good idea (that's why the rubocop rule is here :) )
I think it's better to check that after a save, the expected values are set on the model.

The point is that most of the time specs should not be based on the inner implementation, but on the expected result otherwise the test might be passing even if the what is tested is not working as expected

**Review by vincent-pochet — CHANGES_REQUESTED** [review]

File: db/migrate/20241021140054_create_api_keys.rb (hunk 4 above)
Comment (floganz) [reply db/migrate/20241021140054_create_api_keys.rb:8]: Adding unique index completely make sense. It'd be global rather then scoped by `organization_id` since we need by ApiKey identify only 1 organization.

However, I missed something with @vincent-pochet comment. Index on `organization_id` already in place as part of `t.references`. Maybe you mean some other combination for index?

File: db/migrate/20241021140054_create_api_keys.rb (hunk 5 above)
Comment (floganz) [reply db/migrate/20241021140054_create_api_keys.rb:19]: Yes, that will be removed at some point.

File: spec/models/organization_spec.rb (hunk 11 above)
Comment (floganz) [reply spec/models/organization_spec.rb:146]: Regarding rubocop, there're no warning about stubbing that private method. Maybe we need to check our Rubocop config?

The disabled cop `RSpec/SubjectStub` has a bit different purpose – it ensures that we do not stub thing that we're testing. 
In this case it has false positive match since subject in my case is different, testing that method is called by when another action is done allowed behaviour.

File: db/migrate/20241021140054_create_api_keys.rb (hunk 4 above)
Comment (vincent-pochet) [reply db/migrate/20241021140054_create_api_keys.rb:8]: Ah indeed, forget my remark, the index is here. Sorry for the noise ;)

File: app/models/api_key.rb (hunk 6 above)
Comment (floganz) [reply app/models/api_key.rb:7]: I’ve tried to do add it but it doesn’t look great.

We cannot add `validates :value, presence: true`  with a `before_create` callback that sets the value. The value for new records will be always missing.
If we change callback to `before_validation` we can add a validation but it’s not possible to test, there will no ability to set `nil` value and receive an error that value is missing because of callback. With such setup the validation is a redundancy.

The most reasonable way of adding it by only applying it to update action. But honestly, I'd skip it at all for this particular case.
Let me know if you think I missed something.

**Comment by floganz** [comment]
@ancorcruz @vincent-pochet Thanks for your reviews!
I've pushed changes, can you take another look?

**Review by vincent-pochet — COMMENTED** [review]
Looks great! :rocket: 

Just one comment around the fallback for the lookup of the API key at organization level in the base controller

File: app/controllers/api/base_controller.rb [hunk 12]
```diff
@@ -27,8 +27,8 @@ def authenticate
       true
     end
 
-    def current_organization(api_key = nil)
-      @current_organization ||= Organization.find_by(api_key:)
+    def current_organization(api_key_value = nil)
+      @current_organization ||= ApiKey.find_by(value: api_key_value)&.organization
```
[the remark this answers is not in the document]
Comment (floganz) [reply app/controllers/api/base_controller.rb:31]: Had a chat with Vincent about this. We agreed that it could be done but there’s no big reason for it because with the next step for api key management feature (rotating a key) it might create more confusion then benefits. Since old key shouldn't give access anymore.

**Review by ancorcruz — APPROVED** [review]
LGTM. thanks!
